chalk-ir/src/zip.rs

```diff
@@ -324,21 +452,184 @@ impl<I: Interner> Zip<I> for VariableKind<I> {
+        zipper.zip_substs(
+            None,
+            a.substitution.parameters(interner),
+            b.substitution.parameters(interner),
+        )
+    }
+}
+
+impl<I: Interner> Zip<I> for ProjectionTy<I> {
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
+    where
+        I: 'i,
+    {
+        let interner = zipper.interner();
+        Zip::zip_with(zipper, variance, &a.associated_ty_id, &b.associated_ty_id)?;
+        zipper.zip_substs(
+            None,
+            a.substitution.parameters(interner),
+            b.substitution.parameters(interner),
+        )
+    }
+}
+
+impl<I: Interner> Zip<I> for OpaqueTy<I> {
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
+    where
+        I: 'i,
+    {
+        let interner = zipper.interner();
+        Zip::zip_with(zipper, variance, &a.opaque_ty_id, &b.opaque_ty_id)?;
+        zipper.zip_substs(
+            None,
+            a.substitution.parameters(interner),
+            b.substitution.parameters(interner),
+        )
+    }
+}
+
+impl<I: Interner> Zip<I> for ApplicationTy<I> {
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
     where
         I: 'i,
     {
         let interner = zipper.interner();
-        Zip::zip_with(zipper, a.data(interner), b.data(interner))
+        Zip::zip_with(zipper, variance, &a.name, &b.name)?;
+        match a.name {
+            TypeName::FnDef(fn_def_id) => zipper.zip_substs(
+                Some(zipper.unification_database().fn_def_variance(fn_def_id)),
+                a.substitution.parameters(interner),
+                b.substitution.parameters(interner),
+            ),
+            TypeName::Adt(adt_id) => zipper.zip_substs(
+                Some(zipper.unification_database().adt_variance(adt_id)),
+                a.substitution.parameters(interner),
+                b.substitution.parameters(interner),
+            ),
+            TypeName::Ref(mutbl) => {
+                Zip::zip_with(
+                    zipper,
+                    Variance::Contravariant,
+                    &a.substitution.parameters(interner)[0],
+                    &b.substitution.parameters(interner)[0],
+                )?;
+                let variance = match mutbl {
+                    Mutability::Not => Variance::Covariant,
+                    Mutability::Mut => Variance::Invariant,
+                };
+                Zip::zip_with(
+                    zipper,
+                    variance,
+                    a.substitution.iter(interner).last().unwrap(),
+                    b.substitution.iter(interner).last().unwrap(),
+                )?;
+                Ok(())
+            }
+            _ => zipper.zip_substs(
```

Comment:
Nit: I'd prefer to see this `_` written out explicitly. My usual rubric is "if a new variant is added, how likely is it that it wants custom handling?" I'd say the answer here is "reasonably likely"

Reply:
Oh and I think we want some other bits of variance here...

* [`Ref`](http://rust-lang.github.io/chalk/chalk_ir/enum.TypeName.html#variant.Ref) is contravariant in its lifetime parameter, and either covariant (if Imm) or invariant (if Mutable) in its type parameter
* [`Raw`](http://rust-lang.github.io/chalk/chalk_ir/enum.TypeName.html#variant.Raw) is covariant if const, invariant if mutable

probably more...